### ISSUE TITLE
[Summary] USER_UID/uid related problems (snap/docker)

### ISSUE THREAD

**Issue by wxiaoguang** [go-gitea/gitea#23632]
# Background

1.  Discord: "v1.19.0 installed via snap on Ubuntu server doesn't launch. all is OK with 1.18.5 , but with the last one i got" : `setting.go:322:loadRunModeFrom() [F] Expect user 'gitea' but current user is: root.`
    * "no i didn't change anything, automatic snap update last night to 1.19, break my gitea. i just rollback to 1.18 and all is ok again without touch my config", [the chat link](https://discord.com/channels/322538954119184384/561007778139734027/1087677299375218768)
    * Personally I guess it is likely a misconfiguration problem, but I can't see the details nor get more clues from the reporter at the moment.
2. Docker config: USER_UID=1000 & USER_GID=1000
    * It's unclear what they are used for, or what they affect.
3. #19920
    * Although it was closed, actually it didn't get a fix, it's closed because no time spending on it.
4. The pterodactyl / parkervcp:
    * I used pterodactyl which uses docker with this docker image: ghcr.io/parkervcp/yolks:debian
    * And installs from the config generated from this file: https://github.com/parkervcp/eggs/blob/master/software/gitea/egg-gitea.json
    * Then: `AddPublicKey, calcFingerprintSSHKeygen: 'ssh-keygen -lf /tmp/....' failed with error 'exec(64c52017-2:AddPublicKey) failed: exit status 255(<nil>) stdout:  stderr: No user exists for uid 999
': No user exists for uid 999`

# TODO

I haven't spent time on these problems. Some brief thoughts:

1. If Gitea runs with root, it should be able to setuid/seteuid to switch to the "gitea" user.
    * However, "setuid" correctly in a complex Go program like Gitea is very difficult.
2. Gitea checks the current running user by its name, it's not ideal, because in some rare cases, the UID could be used without an existing user name.
    * Update:  eg: Builtin SSH Server & Docker-rootless only needs a virtual SSH_USER.
3. The USER_UID/USER_GID config for docker-root/docker-rootless are unclear, and it might cause problems if something mismatches or changes.

I guess Gitea needs to spend some time on these problems, work them out, clarify the behaviors and improve documents.

**Comment by lunny** [comment]
> 2.Gitea checks the current running user by its name, it's not ideal, because in some rare cases, the UID could be used without an existing user name.
For the ssh protocol of git, the run user name is also the name which be used in the SSH URL.

**Comment by wxiaoguang** [comment]
> > 2.Gitea checks the current running user by its name, it's not ideal, because in some rare cases, the UID could be used without an existing user name.
> 
> For the ssh protocol of git, the run user name is also the name which be used in the SSH URL.

That's not true for Builtin SSH Server, eg: docker-rootless. Only SSH_USER is necessary, the RUN_USER shouldn't be mixed there.

**Comment by bd4** [comment]
I have this problem with `gitea/gitea:latest-rootless`. No changes to configuration, I have been using `user: "1016:1017"` in the config and it used to work, not it no longer works, I get errors like this:

```
gitea     | 2023/04/26 22:10:29 ...s/setting/setting.go:322:loadRunModeFrom() [F] Expect user 'gitea' but current user is: 
```

**Comment by wxiaoguang** [comment]
Could you help to provide more details? What are the operations before the bug occurs?

Or, you you provide detailed reproducible setup steps? Then I can try to reproduce it on my side.

**Comment by bd4** [comment]
I did not test carefully last time I upgraded, but when I rolled back to `gitea/gitea:1.18-rootless`, the problem goes away. I have a user/group `gitea` on host system, with uid/gid 1016/1017, and my docker-compose looks like this:

```
version: "3"

networks:
  gitea:
    external: false

services:
  server:
    image: gitea/gitea:latest-rootless
    container_name: gitea
    user: "1016:1017"
    restart: unless-stopped
    networks:
      - gitea
    volumes:
      - ./data:/var/lib/gitea
      - ./config:/etc/gitea
      - /etc/timezone:/etc/timezone:ro
      - /etc/localtime:/etc/localtime:ro
    ports:
      - "3001:3000"
      - "2222:2222"
```
Maybe it is enough to just have user: set in docker-compose and upgrade to 1.19? In `config/app.ini`, I have:

```
APP_NAME = Gitea: Git with a cup of tea
RUN_USER = git
RUN_MODE = prod
```

Not sure if RUN_USER has anything to do with it.

**Comment by bd4** [comment]
To be more clear, doing `docker-compose pull` followed by `docker-compose up -d` to upgrade to 1.19 from 1.18 reproduces, and rolling back to 1.18 by changing the image from `gitea/gitea:latest-rootless` to `gitea/gitea:1.18-rootless` fixes it.

**Comment by bd4** [comment]
It is running on Ubuntu 22.04, but no snaps - just have docker and docker-compose installed and manage some docker instances manually.

**Comment by bota87** [comment]
I have probably the same issue as reported in #24385 
I just tried with the latest release `1.19.2-rootless` and the issue persist, if I remove `user: "1016:1017"` it works

**Comment by wxiaoguang** [comment]
I have a question, why did you set `user: "1016:1017"` ?

**Comment by bota87** [comment]
because user 1000 it's already used for another service and I want to keep permissions separated

**Comment by wxiaoguang** [comment]
I'm not sure whether it's related (haven't looked into the problem at the moment): 

Gitea's docker-rootless hard-coded 1000:

https://github.com/go-gitea/gitea/blob/ad03c6e0a36033c6f59262d8cfd6416ae3cc93d6/Dockerfile.rootless#L40-L49

https://github.com/go-gitea/gitea/blob/ad03c6e0a36033c6f59262d8cfd6416ae3cc93d6/Dockerfile.rootless#L62

**Comment by bd4** [comment]
My understanding is that the USER from the dockerfile is the default, but it's it can be overridden by the user config in docker compose or appropriate docker run args: https://docs.docker.com/compose/compose-file/05-services/#user

**Comment by wxiaoguang** [comment]
Yup, but Gitea can't find the user for UID=10xx (non-1000), then bug occurs ..... actually it's also related another problem (#24389), the IsRunUserMatchCurrentUser logic is fragile.

**Comment by bd4** [comment]
Ok so the issue happens when using a host UID that does not exist in the docker image? And it stopped working in 1.19 because the code is trying to determine the user name, and before it wasn't?

**Comment by wxiaoguang** [comment]
Yes

**Comment by wxiaoguang** [comment]
Sorry for bothering, one more question, do you have "DISABLE_SSH=false" and "START_SSH_SERVER=true" in your app.ini , aka, have you enabled Gitea's builtin SSH server?

**Comment by bd4** [comment]
> Sorry for bothering, one more question, do you have "DISABLE_SSH=false" and "START_SSH_SERVER=true" in your app.ini , aka, have you enabled Gitea's builtin SSH server?

Yes I have both those set in the config, so is using the builtin server.

**Comment by wxiaoguang** [comment]
If you could build your own docker image, welcome to try this fix. Or I think it could be merged in one or two days soon then you can try the 1.19 nightly image.

**Comment by wxiaoguang** [comment]
https://hub.docker.com/r/gitea/gitea/tags?page=1&name=1.19-dev  these images contains https://github.com/go-gitea/gitea/pull/24435

**Comment by bota87** [comment]
1.19-dev-rootless isn't updated and didn't worked for me, 1.19-dev-linux-amd64-rootless works

**Comment by wxiaoguang** [comment]
@techknowlogick

**Comment by wxiaoguang** [comment]
> Another thing, I set `ENABLE_SSH_LOG = true` to be able to diagnose ssh problems in the container's sshd, but no sshd errors were written to the log file (which is working for other cases). I wonder if that setting is working?

That's another longstanding problem: difficult to debug SSH part with logs. "ENABLE_SSH_LOG" only sends a few (usually the last error log) to the web server, then the error log will be shown in Gitea server's log with prefix "ssh: ". It's not useful in most cases.

**Comment by tomholford** [comment]
@lonix1 thanks for sharing your findings. I also spent many hours trying to debug this issue. A configurable UID / GID via docker env vars would be really useful

**Comment by tomholford** [comment]
@wxiaoguang @lonix1 @wkrasnicki I started on a potential solution in #27405. Could use your feedback to see if this direction makes sense, and what other changes need to be made to resolve this issue.

**Comment by wxiaoguang** [comment]
Thank you for inviting me to review. At the moment I don't have full understanding for this problem. So feel free to propose a complete solution (with explanation and documents), maybe other users&maintainers could help and suggest.

**Comment by tuomotalvitie** [comment]
After spending some time wondering about this, I've been looking at the difference between
`gitea/docker/root/usr/bin/entrypoint`
and
`gitea/docker/rootless/usr/local/bin/docker-entrypoint.sh`

The first one has sed magic for changing the USER_GID and USER_ID which the second one lacks. If the second one is actually used by rootless, this might have something to do with the issue.

**Comment by maykelvink** [comment]
Ideally you want to overwrite the UID/GID that's being used to start the Gitea binary as environment arguments. You **cannot** add arguments to the Dockerfile if you are not going to build it locally. e.g. I'm going pulling a pre-build image from [dockerhub](https://hub.docker.com/r/gitea/gitea). 

In order to start the binary as a certain user you could add a PUID/PGID as environment variables in docker compose and start the gitea binary for example with S6 Overlay. 

exec s6-setuidgid "$PUID:$PGID" gitea

**Comment by jmptbl** [comment]
Hello! I was surprised to see this issue persisting for so long. The simplest fix would be to just give this Docker image a less common UID and GID, with the expectation that a random UID+GID would be unlikely to clash with an existing user account on the host system. It seems strange to release a public image that uses UID and GID 1000!

Anyhow, the easiest workaround without building a custom image is to follow @maykelvink /etc/passwd bind mounting method. I combined it with an /etc/group bind mount as well, so that I could change the UID and GID. Like this I've gotten SSH pass-through working with the Docker shell method.

**Comment by wxiaoguang** [comment]
No progress, no interest, give up
